Mathlib/Topology/Algebra/OpenSubgroup.lean

```diff
@@ -454,3 +454,107 @@ instance [ContinuousMul G] : Lattice (OpenNormalSubgroup G) :=
+
+open Set Filter
+
+@[to_additive]
+lemma exist_mul_closure_nhd {W : Set G} (WClopen : IsClopen W) : ∃ T ∈ 𝓝 (1 : G), W * T ⊆ W := by
+  apply WClopen.isClosed.isCompact.induction_on (p := fun S ↦ ∃ T ∈ 𝓝 (1 : G), S * T ⊆ W)
+    ⟨Set.univ ,by simp only [univ_mem, empty_mul, empty_subset, and_self]⟩
+    (fun _ _ huv ⟨T, hT, mem⟩ ↦ ⟨T, ⟨hT, (mul_subset_mul_right huv).trans mem⟩⟩)
+    fun U V ⟨T₁, hT₁, mem1⟩ ⟨T₂, hT₂, mem2⟩ ↦ ⟨T₁ ∩ T₂, ⟨inter_mem hT₁ hT₂, by
+      rw [union_mul]
+      exact union_subset (mul_subset_mul_left inter_subset_left |>.trans mem1)
+        (mul_subset_mul_left inter_subset_right |>.trans mem2)⟩⟩
+  intro x memW
+  have : (x, 1) ∈ (fun p ↦ p.1 * p.2) ⁻¹' W := by simp only [mem_preimage, mul_one, memW]
+  rcases isOpen_prod_iff.mp (continuous_mul.isOpen_preimage W <| WClopen.2) x 1 this with
+    ⟨U, V, h1, h2, h3, h4, h5⟩
+  have h6 : U * V ⊆ W := mul_subset_iff.mpr (fun _ hx _ hy ↦ h5 (mk_mem_prod hx hy))
+  exact ⟨U ∩ W, ⟨⟨U, ⟨IsOpen.mem_nhds h1 h3, ⟨W, ⟨fun _ a ↦ a, rfl⟩⟩⟩⟩,
+    ⟨V, ⟨IsOpen.mem_nhds h2 h4, fun _ a ↦ h6 ((mul_subset_mul_right inter_subset_left) a)⟩⟩⟩⟩
+
+@[to_additive]
+lemma exists_mulInvClosureNhd {W : Set G} (WClopen : IsClopen W) :
+    ∃ T, mulInvClosureNhd T W := by
+  rcases exist_mul_closure_nhd WClopen with ⟨S, hS, h⟩
+  rcases mem_nhds_iff.mp hS with ⟨U, h1, h2, h3⟩
+  use U ∩ U⁻¹
+  constructor
+  · exact IsOpen.mem_nhds (h2.inter h2.inv) (by simp only [mem_inter_iff, h3, Set.mem_inv, inv_one,
+    and_self])
+  · simp only [inter_inv, inv_inv, inter_comm]
+  · exact h2.inter h2.inv
+  · exact fun a ha ↦ h (mul_subset_mul_left h1 (mul_subset_mul_left inter_subset_left ha))
+
+@[to_additive]
+theorem existOpenSubgroupSubClopenNhdsOfOne {G : Type*} [Group G] [TopologicalSpace G]
+    [TopologicalGroup G] [CompactSpace G] {W : Set G} (WClopen : IsClopen W) (einW : 1 ∈ W) :
+    ∃ H : OpenSubgroup G, (H : Set G) ⊆ W := by
+  rcases exists_mulInvClosureNhd WClopen with ⟨V, hV⟩
+  let S : Subgroup G := {
+    carrier := ⋃ n , V ^ (n + 1)
+    mul_mem' := fun ha hb ↦ by
+      rcases mem_iUnion.mp ha with ⟨k, hk⟩
+      rcases mem_iUnion.mp hb with ⟨l, hl⟩
+      apply mem_iUnion.mpr
+      use k + 1 + l
+      rw [add_assoc, pow_add]
+      exact Set.mul_mem_mul hk hl
+    one_mem' := by
+      apply mem_iUnion.mpr
+      use 0
+      simp only [zero_add, pow_one, mem_of_mem_nhds hV.nhd]
+    inv_mem' := fun ha ↦ by
+      rcases mem_iUnion.mp ha with ⟨k, hk⟩
+      apply mem_iUnion.mpr
+      use k
+      rw [← hV.inv]
+      simpa only [inv_pow, Set.mem_inv, inv_inv] using hk }
+  have : IsOpen (⋃ n , V ^ (n + 1)) := by
+    refine isOpen_iUnion (fun n ↦ ?_)
+    rw [pow_succ]
+    exact IsOpen.mul_left hV.isOpen
+  use ⟨S, this⟩
+  have mulVpow (n : ℕ) : W * V ^ (n + 1) ⊆ W := by
+    induction' n with n ih
+    · simp only [zero_add, pow_one, hV.mul]
+    · rw [pow_succ, ← mul_assoc]
+      exact le_trans (Set.mul_subset_mul_right ih) <| hV.mul
```

Comment:
What is `<|` doing here? There is no parsing ambiguity without it. And what about writing `exact (Set.mul_subset_mul_right ih).trans hV.mul` instead?